homeassistant/components/nextcloud/__init__.py

```diff
@@ -27,125 +33,164 @@
-    "nextcloud_shares_num_shares_link_no_password",
-    "nextcloud_shares_num_fed_shares_sent",
-    "nextcloud_shares_num_fed_shares_received",
-    "nextcloud_shares_permissions_3_1",
-    "nextcloud_server_webserver",
-    "nextcloud_server_php_version",
-    "nextcloud_server_php_memory_limit",
-    "nextcloud_server_php_max_execution_time",
-    "nextcloud_server_php_upload_max_filesize",
-    "nextcloud_database_type",
-    "nextcloud_database_version",
-    "nextcloud_activeUsers_last5minutes",
-    "nextcloud_activeUsers_last1hour",
-    "nextcloud_activeUsers_last24hours",
-)
 
 
 def setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the Nextcloud integration."""
-    # Fetch Nextcloud Monitor api data
-    conf = config[DOMAIN]
+
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.6.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
+
+    if DOMAIN in config:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=config[DOMAIN],
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the Nextcloud integration."""
+
+    def _connect_nc():
+        return NextcloudMonitor(
+            entry.data[CONF_URL], entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD]
+        )
 
     try:
-        ncm = NextcloudMonitor(conf[CONF_URL], conf[CONF_USERNAME], conf[CONF_PASSWORD])
+        ncm = await hass.async_add_executor_job(_connect_nc)
     except NextcloudMonitorError:
         _LOGGER.error("Nextcloud setup failed - Check configuration")
         return False
 
-    hass.data[DOMAIN] = get_data_points(ncm.data)
-    hass.data[DOMAIN]["instance"] = conf[CONF_URL]
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][
+        entry.data[CONF_URL]
+    ] = coordinator = NextcloudDataUpdateCoordinator(
+        hass,
+        ncm,
+        entry,
+    )
 
-    def nextcloud_update(event_time):
-        """Update data from nextcloud api."""
-        try:
-            ncm.update()
-        except NextcloudMonitorError:
-            _LOGGER.error("Nextcloud update failed")
-            return False
+    await coordinator.async_config_entry_first_refresh()
 
-        hass.data[DOMAIN] = get_data_points(ncm.data)
-        hass.data[DOMAIN]["instance"] = conf[CONF_URL]
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
-    # Update sensors on time interval
-    track_time_interval(hass, nextcloud_update, conf[CONF_SCAN_INTERVAL])
+    return True
 
-    for platform in PLATFORMS:
-        discovery.load_platform(hass, platform, DOMAIN, {}, config)
 
-    return True
+class NextcloudDataUpdateCoordinator(DataUpdateCoordinator):
```

Comment:
Can the coordinator not be implemented in a follow-up PR?
Config-flow PRs are already quite big.

If it's not possible, then I'd recommend moving this to a separate `coordinator` module

Reply:
tbh, i'm not aware of how to get rid of the following block (_which somehow did a similar job as an update coordinator_) without implementing the update coordinator

```python
    # Update sensors on time interval
    track_time_interval(hass, nextcloud_update, conf[CONF_SCAN_INTERVAL])

    for platform in PLATFORMS:
        discovery.load_platform(hass, platform, DOMAIN, {}, config)
```

Reply:
if there is a suitable solution for this problem, then sure i will go with it and do the coordinator in a follow-up PR 👍

Reply:
Maybe the coordinator could be a preliminary PR then, instead of a follow-up PR?